Add a new project template for expo

nickvallee opened the pull request:
Fix #146 

- tested with the newly published `@bonfhir/gluestack-ui` package. Works successfully!
- small alteration on the `runPrettier` function because expo apps do no have a `src` folder, so I added the option to provide the file path to prettier, while keeping the default path.

nickvallee committed:
Add a new project template for expo

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            throw new Error(`Directory ${cwd} already exists`);
+          }
+          await execAsync(
+            `npx create-expo-app ${cwd} -t expo-template-blank-typescript`,
```
This does not work if the user has never installed the `create-expo-app` package.

I advise doing this:


```suggestion
            `npx -y create-expo-app ${cwd} -t expo-template-blank-typescript`,
```

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+      {
+        title: "Create Expo project",
+        task: async ({ options: { cwd, name } }) => {
+          await writeFile(`${cwd}/app.json`, APP_JSON_CONTENT(name), "utf8");
```
This file is created by `create-expo-app`.
It should not be overriden blindly IMHO, as it wiil be very fragile in light of `create-expo-app` tweaks.
We should modify it to add the modifications required.
In this case I believe the only thing changed is `"jsEngine": "jsc",`

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+
+export default function App() {
+  return (
+    <FhirQueryProvider fhirClient={client}>
```
Why isn't the `GluestackUIProvider` configured here?
cf. https://gluestack.io/ui/docs/guides/install-expo

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+  );
+}
+
+const styles = StyleSheet.create({
```
My understanding is that styling should be handled by gluestack-ui.
This is not present in an application created by `npm create gluestack`, so it should probably not be here.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+import PatientComponent from "./PatientComponent";
+
+// Note: if you get [TypeError: Network request failed] errors, you may need to
+// run the following command: ipconfig getifaddr en0
```
I think we need to reword this differently.
This command won't work on Windows, or won't work if the network device in use is not `en0`.
I could not find a reference to that specific command in the GH issue listed.
Also - it worked for me OOB on ios, so I don't know how often this error is that prevalent?
(I got a different error on Android - see global comment.)

nickvallee replied:
it didn't happen at the beginning of working on this, and now it's no longer happening for me. Not sure exactly sure why it happened to trigger the issue, but it was the first time it happened to me ever.

I will remove the comment, as it is likely uncommon and is not relevant to the project.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+import { Text } from "@gluestack-ui/themed";
+
+export default function PatientComponent() {
+  const patientData = useFhirRead(
```
This does not provide a good OOB experience as patient ids are different on all installs.
We should use a [search instead](https://bonfhir.dev/packages/query/queries/use-fhir-search), and get the first one (`patientsSearchQuery.searchMatch()[0]?...`)

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+const TSCONFIG_CONTENT = `{
+    "extends": "expo/tsconfig.base",
+    "compilerOptions": {
+      "moduleResolution": "bundler",
```
Same coment here as for the `app.json` file - we should modify this file, not override it.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+  unstable_enablePackageExports: true,
+};
+
+config.resolver.sourceExts.push('mjs');
```
This could probably be written as a single instruction:


```suggestion
config.resolver = {
  ...config.resolver,
  sourceExts: [
    ...config.resolver.sourceExts,
    "mjs"
  ],
  unstable_enablePackageExports: true,
};
```

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            METRO_CONFIG_CONTENT,
+            "utf8",
+          );
+          await writeFile(
```
Why writing a new `package.json` - I can't identify the difference with the one created by `create-expo-app`.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+    api.cache(true);
+    return {
+      presets: ['babel-preset-expo'],
+      plugins: [
```
Same content as with `app.json` and `tsconfig.json` - we should modify the file to add the missing lines, not override it.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+
+          await packageManager.addDev(
+            cwd,
+            "@babel/core",
```
`@babel/core` is already included in the `create-expo-app` template - we should not diuplicate it as we will end up with wronge version for expo. There is a warning about this already for some packages in the console.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            "@babel/plugin-transform-class-properties",
+            "@babel/plugin-transform-private-methods",
+            "@babel/plugin-transform-private-property-in-object",
+            "@types/react",
```
`@types/react` is already included in the `create-expo-app` template - we should not diuplicate it as we will end up with wronge version for expo. There is a warning about this already for some packages in the console.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            "@babel/plugin-transform-private-property-in-object",
+            "@types/react",
+            "@types/react-native",
+            "typescript",
```
`@babel/core` is already included in the `create-expo-app` template - we should not diuplicate it as we will end up with wronge version for expo. There is a warning about this already for some packages in the console.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            "@bonfhir/gluestack-ui",
+            "@gluestack-style/react",
+            "@gluestack-ui/themed",
+            "expo",
```
`expo` is already included in the `create-expo-app` template - we should not diuplicate it as we will end up with wronge version for expo. There is a warning about this already for some packages in the console.

nickvallee replied:
should be able to remove a number of this packages I add here that come with the creation of the expo app

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            "@gluestack-style/react",
+            "@gluestack-ui/themed",
+            "expo",
+            "expo-status-bar",
```
`expo-status-bar` is already included in the `create-expo-app` template - we should not diuplicate it as we will end up with wronge version for expo. There is a warning about this already for some packages in the console.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            "@gluestack-ui/themed",
+            "expo",
+            "expo-status-bar",
+            "react",
```
`react` is already included in the `create-expo-app` template - we should not diuplicate it as we will end up with wronge version for expo. There is a warning about this already for some packages in the console.

nickvallee commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+          );
+          await writeFile(
+            `${cwd}/package.json`,
+            PACKAGE_JSON_CONTENT(name),
```
I realize that some of these files will have been created with the `create-expo cli` command, so I can remove the writing of this package.json file, and some of the packages I add below that come with the creation of the expo app.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            "expo",
+            "expo-status-bar",
+            "react",
+            "react-native",
```
`react-native` is already included in the `create-expo-app` template - we should not diuplicate it as we will end up with wronge version for expo. There is a warning about this already for some packages in the console.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            "expo-status-bar",
+            "react",
+            "react-native",
+            "react-native-svg",
```
`react-native-svg` should be installed using [`expo install...`](https://docs.expo.dev/more/expo-cli/#install) and not `npm install` to ensure the proper version is used.
This is crucial for expo workflows.

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,277 @@
+            "@bonfhir/react",
+            "@bonfhir/gluestack-ui",
+            "@gluestack-style/react",
+            "@gluestack-ui/themed",
```
The glustack template also installs `@gluestack-ui/config` and `@legendapp/motion` - we should probably add them here.

julienblin reviewed (changes requested):
Thank you - this is a good start!

However I feel like there is still some work to do.
Aside from the comments I left, I also want to add 2 points:

- The app did not work on the Android emulator on my machine out-of-the-box; I did not investigate why, but here is the screenshot:
<img width="494" alt="image" src="https://github.com/bonfhir/bonfhir/assets/123453/d63c3c44-169a-4455-b2a4-752a51d8e78b">

- We install prettier and add a `.prettierignore` file, but no scripts to run it in the project. Please take a look at other templates and apply the same pattern.

Finally we have to ask ourselves if we want to support expo web in the template.  
Since we have a scripts for it (`web`), I feel we should, and as such should apply the proper `gluestack-ui` recipe for it to work (e.g. at least install `@expo/html-elements`)

nickvallee commented:
@julienblin I believe I have addressed all your comments:
- modified config files and package json instead of overriding
- add -y to create-expo
- add glue stack provider
- removed un-needed dependencies adding
- add missing glue stack dependencies
- use search instead of read to remove use of patient ID
- create prettier script command
- look into android bug and file new issue #213 
- allow support for running project in web
- remove useless CSS

Note on web support: you mentioned `@expo/html-elements` and a gluestack recipe for compatibility with web, but I could not find an exact recipe. There is also documentation on supporting web or expo: https://docs.expo.dev/workflow/web/
I followed that guide, and the project works in the web browser. I was unsure how `@expo/html-element` related to the official docs, so I have left it for now, but can add it if you would like with additional info of why.

If the expo docs instructions are not right, can just give me a bit more information on the glue stack specific recipe, that would be great. all I found was a mention of react-native-web in the docs, is added in the expo docs as well. Maybe I missed it.

julienblin commented on .changeset/red-ears-ring.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@bonfhir/cli": minor
```
We have a special case here when we modify templates as we need to release both `@bonfhir/cli` and `create-bonfhir` at the same time, so we need to add it here.

```suggestion
"@bonfhir/cli": minor
"create-bonfhir": minor
```

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,255 @@
+            await writeFile(appJsonFilePath, updatedConfig, "utf8");
+          } catch (error) {
+            console.error("Error reading or writing the app.json file:", error);
+          }
```
Please refactor into a utility function for modifying json files

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,255 @@
+            cwd,
+            "@babel/plugin-transform-class-properties",
+            "@babel/plugin-transform-private-methods",
+            "@babel/plugin-transform-private-property-in-object",
```
My testing showed that these dependencies are unnecessary as they are already pulled by expo

julienblin commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,255 @@
+              plugins: [
+                ["@babel/plugin-transform-class-properties", { "loose": true }],
+                ["@babel/plugin-transform-private-methods", { "loose": true }],
+                ["@babel/plugin-transform-private-property-in-object", { "loose": true }]
```
My own testing showed that only the `@babel/plugin-transform-private-methods` felt necessary:

```suggestion
                ["@babel/plugin-transform-private-methods", { "loose": true }]
```

julienblin reviewed (changes requested):
Thank you - this is really good - we are close!

The gluestack "recipe" I was referring to was when I executed [`npm create gluestack`](https://gluestack.io/ui/docs/getting-started/installation). We don't necessarily need to follow it, but I was wondering if it was useful.

Here are the last remaining things that I would like you to do:
 - address the comments that I added here - they are all trivial
 - I think we should finish the process by running a `git commit -a . -m "Setup bonFHIR"` command at the end; as it stands, it feels weird that there are pending modifications when opening the project; since expo already creates an initial commit I believe it is fine for us to do the same.

nickvallee commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,237 @@
+        title: "Commit bonFHIR setup changes",
+        task: async ({ options: { cwd } }) => {
+          await execAsync(`git add .`, { cwd });
+          await execAsync(`git commit -m "Setup bonFHIR"`, { cwd });
```
`- a` only commits already tracked files, since we add new files, I slip the command in two

nickvallee commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,237 @@
+        task: async ({ options: { cwd } }) => {
+          const appJsonFilePath = `${cwd}/app.json`;
+          await modifyJsonFile(appJsonFilePath, (config) => {
+            // @ts-expect-error structure of config file subject to change
```
didn't seem worth it to try to resolve this type error (config type of unknown). Wasn't sure the best approach to do that without a lot of extra code. Let me know if you can think of a good solution.

julienblin replied:
I am not a fan of this, although I agree it is not worth typing the config files.
I recommend that you apply my code suggestion in the `modifyJsonFile` method and remove the `@ts-expect-error`.

nickvallee commented on packages/cli/src/templates/expo.ts:
```diff
@@ -0,0 +1,237 @@
+          });
+
+          try {
+            const babelConfigFilePath = `${cwd}/babel.config.js`;
```
note that babel config is not a json file, so we are not using the function.

nickvallee commented:
@julienblin thank you for the thorough testing. I have updated with the changes. I will work to be more thorough in the future.

- modify json util function
- update change set
- remove useless dependencies
- add new commit to end of script

julienblin commented on packages/cli/src/templates/utils/modify-json-file.ts:
```diff
@@ -0,0 +1,16 @@
+import { readFile, writeFile } from "node:fs/promises";
+
+export async function modifyJsonFile<T>(
```
This is good - but to avoid type errors I would recommend defaulting the `T` to `Record<string, any>`


```suggestion
export async function modifyJsonFile<T = Record<string, any>>(
```

nickvallee replied:
ahhhh there it is! thanks

julienblin reviewed (approved):
This is great!
Just have a small recommendation for the JSON type errors.
Let's add that in!!

julienblin reviewed (approved):
This is great! 🎉
Just have a small recommendation for the JSON type errors.
Let's add that in!!

nickvallee commented:
@julienblin should all be good! I would like to merge but it says I can't because there are unresolved conversations, but when I click on conversations they no longer exist. Might be a bug in the github UI having to do with rebases and force pushes.

You might have to force merge this one 🤔

julienblin commented:
@nickvallee - yep, weird; GH got confused with the force push.
I force merged it anyway.

Thanks a lot! This is the beginning of new possibilities for bonFHIR!
